### PR TITLE
Fix APCs 'swimming' over bridges

### LINKED ISSUE #14592: misplaced BridgePlaceholder on Tread Lightly (bridge glitch)
flying

I guess a rather unimportend glitch for now but i didn't want to hold it back.

Replay
[OpenRA-2017-12-28T010539Z.zip](https://github.com/OpenRA/OpenRA/files/1590180/OpenRA-2017-12-28T010539Z.zip)

### PR DESCRIPTION
**Pull request by fruestueck**
This is my first change on the OpenRA code. Constructive feedback appreciated : )

Depends on PR #14602 
Fixes #14592 

bridge2

ToDo: Get rid of 'splashes' and set the shadow right.
I don't know how to approach the fix for that. Possibly i missed something.
Surely i didn't fix it proper so, open for feedback : )

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnTerrain.cs [hunk 1]
```diff
@@ -55,10 +55,11 @@ void ITick.Tick(Actor self)
 			var wantsGranted = info.TerrainTypes.Contains(currentTerrain);
 			if (currentTerrain != previousTerrain)
 			{
-				if (wantsGranted && conditionToken == ConditionManager.InvalidConditionToken)
-					conditionToken = conditionManager.GrantCondition(self, info.Condition);
-				else if (!wantsGranted && conditionToken != ConditionManager.InvalidConditionToken)
-					conditionToken = conditionManager.RevokeCondition(self, conditionToken);
+				if (self.Location.Layer == 0)
```
Comment (fruestueck) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnTerrain.cs]: Maby the Layer check is only needed for the GrandCondition part.

Comment (pchote) [reply OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnTerrain.cs]: This should move up into the `wantsGranted` definition.

**Comment by fruestueck** [comment]
Squashed

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnTerrain.cs [hunk 2]
```diff
@@ -52,7 +52,7 @@ void ITick.Tick(Actor self)
 				return;
 
 			var currentTerrain = self.World.Map.GetTerrainInfo(self.Location).Type;
-			var wantsGranted = info.TerrainTypes.Contains(currentTerrain);
+			var wantsGranted = info.TerrainTypes.Contains(currentTerrain) && (self.Location.Layer == 0);
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnTerrain.cs]: Can you remove the parentheses, please?

Comment (fruestueck) [reply OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnTerrain.cs]: Done

File: OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnTerrain.cs (hunk 1 above)
Comment (fruestueck) [reply OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnTerrain.cs]: Done

**Comment by fruestueck** [comment]
Tried to change the Shadow Offset of an Actor while on bridge.

Added the following code in
`OpenRA.Mods.Common/Traits/Renderer/WithShadows.cs`

```
		IEnumerable<IRenderable> IRenderModifier.ModifyRender(Actor self, ...)
		{
			...
			// Contrails shouldn't cast shadows
			var height = self.World.Map.DistanceAboveTerrain(self.CenterPosition).Length;
--->			if (self.Location.Layer != 0) height = 100;
			var shadowSprites = r.Where(s => !s.IsDecoration)
				.Select(a => a.WithPalette(wr.Palette(info.Palette))
					.OffsetBy(info.Offset - new WVec(0, 0, height))
					.WithZOffset(a.ZOffset + (height + info.ZOffset))
					.AsDecoration());
...
```
Had no effect, what am i doing wrong?

**Comment by pchote** [comment]
Voxel shadows are handled deep inside the rendering code.  You really don't want to get into that here.

**Comment by fruestueck** [comment]
Okay, then i'll stop here. Was a bit too much for the beginning anyway : )
Should i open a nother issue with the Shadow offset and 'swimming' waves so this can be closed on succesful PR?

**Comment by pchote** [comment]
#14592 has been merged, so this needs a rebase to remove the changes from the squashed commit.

**Comment by pchote** [comment]
Feel free to add the wake fix here.

**Comment by fruestueck** [comment]
new PR opend

**Pull request closed without merging**